// backendless/es/messaging/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _dec, _dec2, _dec3, _dec4, _dec5, _dec6, _dec7, _dec8, _desc, _value, _obj, _init, _init2, _init3, _init4, _init5, _init6, _init7, _init8;

var _utils = require('../utils');

var _utils2 = _interopRequireDefault(_utils);

var _decorators = require('../decorators');

var _publishOptionsHeader = require('./publish-options-header');

var _publishOptionsHeader2 = _interopRequireDefault(_publishOptionsHeader);

var _publishOptions = require('./publish-options');

var _publishOptions2 = _interopRequireDefault(_publishOptions);

var _deliveryOptions = require('./delivery-options');

var _deliveryOptions2 = _interopRequireDefault(_deliveryOptions);

var _bodyParts = require('./body-parts');

var _bodyParts2 = _interopRequireDefault(_bodyParts);

var _subscriptionsOptions = require('./subscriptions-options');

var _subscriptionsOptions2 = _interopRequireDefault(_subscriptionsOptions);

var _subscribe = require('./subscribe');

var _publish = require('./publish');

var _push = require('./push');

var _pushWithTemplate = require('./push-with-template');

var _sendEmail = require('./send-email');

var _cancel = require('./cancel');

var _registerDevice = require('./register-device');

var _unregisterDevice = require('./unregister-device');

var _getRegistration = require('./get-registration');

var _getMessageStatus = require('./get-message-status');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _applyDecoratedDescriptor(target, property, decorators, descriptor, context) {
  var desc = {};
  Object['ke' + 'ys'](descriptor).forEach(function (key) {
    desc[key] = descriptor[key];
  });
  desc.enumerable = !!desc.enumerable;
  desc.configurable = !!desc.configurable;

  if ('value' in desc || desc.initializer) {
    desc.writable = true;
  }

  desc = decorators.slice().reverse().reduce(function (desc, decorator) {
    return decorator(target, property, desc) || desc;
  }, desc);

  if (context && desc.initializer !== void 0) {
    desc.value = desc.initializer ? desc.initializer.call(context) : void 0;
    desc.initializer = undefined;
  }

  if (desc.initializer === void 0) {
    Object['define' + 'Property'](target, property, desc);
    desc = null;
  }

  return desc;
}

var Messaging = (_dec = (0, _decorators.deprecated)('Backendless.Messaging', 'Backendless.Messaging.subscribe'), _dec2 = (0, _decorators.deprecated)('Backendless.Messaging', 'Backendless.Messaging.publish'), _dec3 = (0, _decorators.deprecated)('Backendless.Messaging', 'Backendless.Messaging.sendEmail'), _dec4 = (0, _decorators.deprecated)('Backendless.Messaging', 'Backendless.Messaging.cancel'), _dec5 = (0, _decorators.deprecated)('Backendless.Messaging', 'Backendless.Messaging.registerDevice'), _dec6 = (0, _decorators.deprecated)('Backendless.Messaging', 'Backendless.Messaging.getRegistrations'), _dec7 = (0, _decorators.deprecated)('Backendless.Messaging', 'Backendless.Messaging.unregisterDevice'), _dec8 = (0, _decorators.deprecated)('Backendless.Messaging', 'Backendless.Messaging.getMessageStatus'), (_obj = {

  Bodyparts: _bodyParts2.default,
  PublishOptions: _publishOptions2.default,
  DeliveryOptions: _deliveryOptions2.default,
  SubscriptionOptions: _subscriptionsOptions2.default,
  PublishOptionsHeaders: _publishOptionsHeader2.default,

  subscribeSync: _utils2.default.synchronized(_subscribe.subscribe),
  subscribe: _utils2.default.promisified(_subscribe.subscribe),

  publishSync: _utils2.default.synchronized(_publish.publish),
  publish: _utils2.default.promisified(_publish.publish),

  push: _utils2.default.promisified(_push.push),

  pushWithTemplate: _utils2.default.promisified(_pushWithTemplate.pushWithTemplate),

  sendEmailSync: _utils2.default.synchronized(_sendEmail.sendEmail),
  sendEmail: _utils2.default.promisified(_sendEmail.sendEmail),

  cancelSync: _utils2.default.synchronized(_cancel.cancel),
  cancel: _utils2.default.promisified(_cancel.cancel),

  registerDeviceSync: _utils2.default.synchronized(_registerDevice.registerDevice),
  registerDevice: _utils2.default.promisified(_registerDevice.registerDevice),

  getRegistrationsSync: _utils2.default.synchronized(_getRegistration.getRegistrations),
  getRegistrations: _utils2.default.promisified(_getRegistration.getRegistrations),

  unregisterDeviceSync: _utils2.default.synchronized(_unregisterDevice.unregisterDevice),
  unregisterDevice: _utils2.default.promisified(_unregisterDevice.unregisterDevice),

  getMessageStatusSync: _utils2.default.synchronized(_getMessageStatus.getMessageStatus),
  getMessageStatus: _utils2.default.promisified(_getMessageStatus.getMessageStatus)

}, (_applyDecoratedDescriptor(_obj, 'subscribeSync', [_dec], (_init = Object.getOwnPropertyDescriptor(_obj, 'subscribeSync'), _init = _init ? _init.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'publishSync', [_dec2], (_init2 = Object.getOwnPropertyDescriptor(_obj, 'publishSync'), _init2 = _init2 ? _init2.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init2;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'sendEmailSync', [_dec3], (_init3 = Object.getOwnPropertyDescriptor(_obj, 'sendEmailSync'), _init3 = _init3 ? _init3.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init3;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'cancelSync', [_dec4], (_init4 = Object.getOwnPropertyDescriptor(_obj, 'cancelSync'), _init4 = _init4 ? _init4.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init4;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'registerDeviceSync', [_dec5], (_init5 = Object.getOwnPropertyDescriptor(_obj, 'registerDeviceSync'), _init5 = _init5 ? _init5.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init5;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'getRegistrationsSync', [_dec6], (_init6 = Object.getOwnPropertyDescriptor(_obj, 'getRegistrationsSync'), _init6 = _init6 ? _init6.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init6;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'unregisterDeviceSync', [_dec7], (_init7 = Object.getOwnPropertyDescriptor(_obj, 'unregisterDeviceSync'), _init7 = _init7 ? _init7.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init7;
  }
}), _obj), _applyDecoratedDescriptor(_obj, 'getMessageStatusSync', [_dec8], (_init8 = Object.getOwnPropertyDescriptor(_obj, 'getMessageStatusSync'), _init8 = _init8 ? _init8.value : undefined, {
  enumerable: true,
  configurable: true,
  writable: true,
  initializer: function initializer() {
    return _init8;
  }
}), _obj)), _obj));

exports.default = Messaging;
