// backendless/es/messaging/polling-proxy.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _request = require('../request');

var _request2 = _interopRequireDefault(_request);

var _async = require('../request/async');

var _async2 = _interopRequireDefault(_async);

var _proxy = require('./proxy');

var _proxy2 = _interopRequireDefault(_proxy);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var PollingProxy = function (_Proxy) {
  _inherits(PollingProxy, _Proxy);

  function PollingProxy(url) {
    _classCallCheck(this, PollingProxy);

    var _this = _possibleConstructorReturn(this, (PollingProxy.__proto__ || Object.getPrototypeOf(PollingProxy)).call(this));

    _this.restUrl = url;
    _this.timer = 0;
    _this.timeout = 0;
    _this.interval = 1000;
    _this.stopped = false;
    _this.responder = new _async2.default(_this.onMessage, _this.onError, _this);

    _this.poll();
    return _this;
  }

  _createClass(PollingProxy, [{
    key: 'poll',
    value: function poll() {
      var _this2 = this;

      clearTimeout(this.timer);
      clearTimeout(this.timeout);

      this.timeout = setTimeout(function () {
        return _this2.onTimeout();
      }, 30 * 1000);

      _request2.default.get({
        url: this.restUrl,
        isAsync: true,
        asyncHandler: this.responder
      });
    }
  }, {
    key: 'close',
    value: function close() {
      clearTimeout(this.timer);
      clearTimeout(this.timeout);

      this.stopped = true;
    }
  }, {
    key: 'onTimeout',
    value: function onTimeout() {
      clearTimeout(this.timeout);

      this.poll();
    }
  }, {
    key: 'onMessage',
    value: function onMessage(data) {
      var _this3 = this;

      clearTimeout(this.timer);
      clearTimeout(this.timeout);

      if (!this.stopped) {
        this.timer = setTimeout(function () {
          return _this3.poll();
        }, this.interval);

        this.fireEvent('messageReceived', data);
      }
    }
  }, {
    key: 'onError',
    value: function onError() {
      var _this4 = this;

      clearTimeout(this.timer);
      clearTimeout(this.timeout);

      if (!this.stopped) {
        this.timer = setTimeout(function () {
          return _this4.poll();
        }, this.interval);
      }
    }
  }]);

  return PollingProxy;
}(_proxy2.default);

exports.default = PollingProxy;
