// backendless/es/messaging/socket-proxy.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _proxy = require('./proxy');

var _proxy2 = _interopRequireDefault(_proxy);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var SocketProxy = function (_Proxy) {
  _inherits(SocketProxy, _Proxy);

  function SocketProxy(url) {
    _classCallCheck(this, SocketProxy);

    var _this = _possibleConstructorReturn(this, (SocketProxy.__proto__ || Object.getPrototypeOf(SocketProxy)).call(this));

    _this.reconnectWithPolling = true;

    try {
      var socket = _this.socket = new WebSocket(url);

      socket.onopen = function () {
        return _this.sockOpen();
      };
      socket.onerror = function (error) {
        return _this.sockError(error);
      };
      socket.onclose = function () {
        return _this.onSocketClose();
      };
      socket.onmessage = function (event) {
        return _this.onMessage(event);
      };
    } catch (e) {
      setTimeout(function () {
        return _this.onSocketClose();
      }, 100);
    }
    return _this;
  }

  _createClass(SocketProxy, [{
    key: 'onMessage',
    value: function onMessage() {
      this.fireEvent('messageReceived', data);
    }
  }, {
    key: 'onSocketClose',
    value: function onSocketClose(data) {
      if (this.reconnectWithPolling) {
        this.fireEvent('socketClose', data);
      }
    }
  }, {
    key: 'close',
    value: function close() {
      this.reconnectWithPolling = false;

      this.socket.close();
    }
  }]);

  return SocketProxy;
}(_proxy2.default);

exports.default = SocketProxy;
