// backendless/es/messaging/subscribe.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.subscribe = subscribe;

var _urls = require('../urls');

var _urls2 = _interopRequireDefault(_urls);

var _async = require('../request/async');

var _async2 = _interopRequireDefault(_async);

var _subscription = require('./subscription');

var _subscription2 = _interopRequireDefault(_subscription);

var _getChannelProperty = require('./get-channel-property');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function subscribe(channelName, subscriptionCallback, subscriptionOptions, asyncHandler) {
  if (asyncHandler) {
    var callback = new _async2.default(function (props) {
      asyncHandler.success(new _subscription2.default({
        channelName: channelName,
        options: subscriptionOptions,
        channelProperties: props,
        responder: subscriptionCallback,
        restUrl: _urls2.default.messaging(),
        onSubscribe: asyncHandler
      }));
    }, function (data) {
      asyncHandler.fault(data);
    });

    (0, _getChannelProperty.getChannelProperties)(channelName, callback);
  } else {
    var props = (0, _getChannelProperty.getChannelProperties)(channelName);

    return new _subscription2.default({
      channelName: channelName,
      options: subscriptionOptions,
      channelProperties: props,
      responder: subscriptionCallback,
      restUrl: _urls2.default.messaging()
    });
  }
}
