// backendless/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _backendlessRequest = require('backendless-request');

var _backendlessRequest2 = _interopRequireDefault(_backendlessRequest);

var _logging = require('./logging');

var _logging2 = _interopRequireDefault(_logging);

var _counters = require('./counters');

var _counters2 = _interopRequireDefault(_counters);

var _cache = require('./cache');

var _cache2 = _interopRequireDefault(_cache);

var _commerce = require('./commerce');

var _commerce2 = _interopRequireDefault(_commerce);

var _users = require('./users');

var _users2 = _interopRequireDefault(_users);

var _user = require('./users/user');

var _user2 = _interopRequireDefault(_user);

var _customServices = require('./bl/custom-services');

var _customServices2 = _interopRequireDefault(_customServices);

var _events = require('./bl/events');

var _events2 = _interopRequireDefault(_events);

var _geo = require('./geo');

var _geo2 = _interopRequireDefault(_geo);

var _data = require('./data');

var _data2 = _interopRequireDefault(_data);

var _messaging = require('./messaging');

var _messaging2 = _interopRequireDefault(_messaging);

var _device = require('./device');

var _device2 = _interopRequireDefault(_device);

var _files = require('./files');

var _files2 = _interopRequireDefault(_files);

var _localCache = require('./local-cache');

var _localCache2 = _interopRequireDefault(_localCache);

var _localVars = require('./local-vars');

var _localVars2 = _interopRequireDefault(_localVars);

var _initApp = require('./init-app');

var _userAgent = require('./user-agent');

var _currentUser = require('./users/current-user');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var root = (typeof self === 'undefined' ? 'undefined' : _typeof(self)) === 'object' && self.self === self && self || (typeof global === 'undefined' ? 'undefined' : _typeof(global)) === 'object' && global.global === global && global;

var previousBackendless = root && root.Backendless;

var Backendless = {

  get debugMode() {
    return _localVars2.default.debugMode;
  },

  set debugMode(debugMode) {
    _localVars2.default.debugMode = !!debugMode;
  },

  get serverURL() {
    return _localVars2.default.serverURL;
  },

  set serverURL(serverURL) {
    _localVars2.default.serverURL = serverURL;
  },

  get XMLHttpRequest() {
    return _localVars2.default.XMLHttpRequest;
  },

  set XMLHttpRequest(XMLHttpRequest) {
    _localVars2.default.XMLHttpRequest = XMLHttpRequest;
  },

  get applicationId() {
    return _localVars2.default.applicationId;
  },

  get secretKey() {
    return _localVars2.default.secretKey;
  },

  get appPath() {
    return _localVars2.default.appPath;
  },

  initApp: _initApp.initApp,

  getCurrentUserToken: _currentUser.getCurrentUserToken,

  setupDevice: _device2.default.setup,

  browser: (0, _userAgent.getUserAgent)(),

  Request: _backendlessRequest2.default,

  noConflict: function noConflict() {
    if (root) {
      root.Backendless = previousBackendless;
    }

    return Backendless;
  },


  ///-------------------------------------///
  ///-------------- SERVICES -------------///

  Logging: _logging2.default,
  Counters: _counters2.default,
  Cache: _cache2.default,
  Commerce: _commerce2.default,
  Users: _users2.default,
  User: _user2.default,
  CustomServices: _customServices2.default,
  Events: _events2.default,
  Geo: _geo2.default,
  Data: _data2.default,
  Messaging: _messaging2.default,
  Files: _files2.default,

  ///-------------- SERVICES -------------///
  ///-------------------------------------///

  ///-------------------------------------///
  ///--------BACKWARD COMPATIBILITY-------///

  //TODO: do we need to remove it?
  UserService: _users2.default,
  GeoQuery: _geo2.default.Query,
  GeoPoint: _geo2.default.Point,
  GeoCluster: _geo2.default.Cluster,
  Persistence: _data2.default,
  DataQueryBuilder: _data2.default.QueryBuilder,
  LoadRelationsQueryBuilder: _data2.default.LoadRelationsQueryBuilder,
  Bodyparts: _messaging2.default.Bodyparts,
  PublishOptions: _messaging2.default.PublishOptions,
  DeliveryOptions: _messaging2.default.DeliveryOptions,
  SubscriptionOptions: _messaging2.default.SubscriptionOptions,
  PublishOptionsHeaders: _messaging2.default.PublishOptionsHeaders,

  LocalCache: _localCache2.default
  ///--------BACKWARD COMPATIBILITY-------///
  ///-------------------------------------///
};

if (root) {
  root.Backendless = Backendless;
}

exports.default = Backendless;


module.exports = Backendless;
