// backendless/lib/messaging/publish.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.publish = publish;

var _utils = require('../utils');

var _utils2 = _interopRequireDefault(_utils);

var _urls = require('../urls');

var _urls2 = _interopRequireDefault(_urls);

var _request = require('../request');

var _request2 = _interopRequireDefault(_request);

var _deliveryOptions = require('./delivery-options');

var _deliveryOptions2 = _interopRequireDefault(_deliveryOptions);

var _publishOptions = require('./publish-options');

var _publishOptions2 = _interopRequireDefault(_publishOptions);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function publish(channelName, message, publishOptions, deliveryTarget /**, async */) {
  var responder = _utils2.default.extractResponder(arguments);
  var isAsync = !!responder;

  var data = {
    message: message
  };

  if (publishOptions && publishOptions !== responder) {
    if (!(publishOptions instanceof _publishOptions2.default)) {
      throw new Error('Use PublishOption as publishOptions argument');
    }

    _utils2.default.deepExtend(data, publishOptions);
  }

  if (deliveryTarget && deliveryTarget !== responder) {
    if (!(deliveryTarget instanceof _deliveryOptions2.default)) {
      throw new Error('Use DeliveryOptions as deliveryTarget argument');
    }

    _utils2.default.deepExtend(data, deliveryTarget);
  }

  return _request2.default.post({
    url: _urls2.default.messagingChannel(channelName),
    isAsync: isAsync,
    asyncHandler: responder,
    data: data
  });
}
