// backendless/lib/messaging/subscription.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _utils = require('../utils');

var _utils2 = _interopRequireDefault(_utils);

var _async2 = require('../request/async');

var _async3 = _interopRequireDefault(_async2);

var _request = require('../request');

var _request2 = _interopRequireDefault(_request);

var _pollingProxy = require('./polling-proxy');

var _pollingProxy2 = _interopRequireDefault(_pollingProxy);

var _socketProxy = require('./socket-proxy');

var _socketProxy2 = _interopRequireDefault(_socketProxy);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Subscription = function () {
  function Subscription(config) {
    _classCallCheck(this, Subscription);

    this.channelName = config.channelName;
    this.options = config.options;
    this.channelProperties = config.channelProperties;
    this.subscriptionId = null;
    this.restUrl = config.restUrl + '/' + config.channelName;
    this.responder = config.responder;
    this._subscribe(config.onSubscribe);
  }

  _createClass(Subscription, [{
    key: '_subscribe',
    value: function _subscribe() /** async */{
      var responder = _utils2.default.extractResponder(arguments);
      var isAsync = !!responder;
      var self = this;

      var _async = new _async3.default(function (data) {
        self.subscriptionId = data.subscriptionId;
        self._startSubscription();
      }, function (e) {
        responder.fault(e);
      });

      var subscription = _request2.default.post({
        url: this.restUrl + '/subscribe',
        isAsync: isAsync,
        data: this.options,
        asyncHandler: _async
      });

      if (!isAsync) {
        this.subscriptionId = subscription.subscriptionId;
        this._startSubscription();
      }
    }
  }, {
    key: '_startSubscription',
    value: function _startSubscription() {
      var _this = this;

      if (WebSocket) {
        var url = this.channelProperties['websocket'] + '/' + this.subscriptionId;
        this.proxy = new _socketProxy2.default(url);

        this.proxy.on('socketClose', function () {
          _this._switchToPolling();
        });

        this.proxy.on('messageReceived', function () {
          if (_this.responder) {
            _this.responder();
          }
        });
      } else {
        this._switchToPolling();
      }

      this._startSubscription = function () {};
    }
  }, {
    key: 'cancelSubscription',
    value: function cancelSubscription() {
      this.proxy && this.proxy.close();
      this._startSubscription = function () {};
    }
  }, {
    key: '_switchToPolling',
    value: function _switchToPolling() {
      var url = this.restUrl + '/' + this.subscriptionId;
      this.proxy = new _pollingProxy2.default(url);
      var self = this;

      this.proxy.on('messageReceived', function (data) {
        if (data.messages.length) {
          self.responder(data);
        }
      });
    }
  }]);

  return Subscription;
}();

exports.default = Subscription;
